webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -313,109 +294,112 @@ public async Task<string> ExtractAudienceAsync(SKContext context)
-        await this.UpdateResponseStatusOnClient(chatId, "Acquiring external information from planner");
+        // Acquire external information from planner
+        await this.UpdateBotResponseStatusOnClient(chatId, "Acquiring external information from planner");
         var externalInformationTokenLimit = (int)(remainingToken * this._promptOptions.ExternalInformationContextWeight);
         var planResult = await this.AcquireExternalInformationAsync(chatContext, userIntent, externalInformationTokenLimit);
         if (chatContext.ErrorOccurred)
         {
-            return string.Empty;
+            return null;
         }
 
         // If plan is suggested, send back to user for approval before running
-        if (this._externalInformationSkill.ProposedPlan != null)
+        var proposedPlan = this._externalInformationSkill.ProposedPlan;
+        if (proposedPlan != null)
         {
-            chatContext.Variables.Set("prompt", this._externalInformationSkill.ProposedPlan.Plan.Description);
-            return JsonSerializer.Serialize<ProposedPlan>(this._externalInformationSkill.ProposedPlan);
-        }
+            var prompt = proposedPlan.Plan.Description;
+            chatContext.Variables.Set("prompt", prompt);
 
-        // 4. Query relevant semantic memories
-        await this.UpdateResponseStatusOnClient(chatId, "Querying semantic memories");
-        var chatMemoriesTokenLimit = (int)(remainingToken * this._promptOptions.MemoriesResponseContextWeight);
-        var chatMemories = await this._semanticChatMemorySkill.QueryMemoriesAsync(userIntent, chatId, chatMemoriesTokenLimit, chatContext.Memory);
-        if (chatContext.ErrorOccurred)
-        {
-            return string.Empty;
+            // Save a new response to the chat history with the proposed plan content
+            return await this.SaveNewResponseAsync(JsonSerializer.Serialize<ProposedPlan>(proposedPlan), prompt, chatId, userId);
         }
 
-        // 5. Query relevant document memories
-        await this.UpdateResponseStatusOnClient(chatId, "Querying document memories");
+        // Query relevant semantic and document memories
+        await this.UpdateBotResponseStatusOnClient(chatId, "Extracting semantic and document memories");
+        var chatMemoriesTokenLimit = (int)(remainingToken * this._promptOptions.MemoriesResponseContextWeight);
         var documentContextTokenLimit = (int)(remainingToken * this._promptOptions.DocumentContextWeight);
-        var documentMemories = await this._documentMemorySkill.QueryDocumentsAsync(userIntent, chatId, documentContextTokenLimit, chatContext.Memory);
+        tasks = await Task.WhenAll<string>(
+            this._semanticChatMemorySkill.QueryMemoriesAsync(userIntent, chatId, chatMemoriesTokenLimit, chatContext.Memory),
+            this._documentMemorySkill.QueryDocumentsAsync(userIntent, chatId, documentContextTokenLimit, chatContext.Memory)
+        );
+
         if (chatContext.ErrorOccurred)
```

Comment:
This is not your code but we probably don't need this anymore as we are not passing the context to the `QueryXXXAsync` methods anymore. We need to throw and catch exceptions instead.

Reply:
Can be addressed in a separate PR.

Reply:
fixed